samples/DatabaseContainers/DatabaseContainers.ApiService/ApiEndpoints.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Dapper;
+using Microsoft.Data.SqlClient;
+using MySqlConnector;
+using Npgsql;
+
+namespace DatabaseContainers.ApiService;
+
+public static class ApiEndpoints
+{
+    public static WebApplication MapTodosApi(this WebApplication app)
+    {
+        app.MapGet("/todos", async (NpgsqlDataSource db) =>
+        {
+            const string sql = """
+                SELECT Id, Title, IsComplete
+                FROM Todos
+                """;
+            var connection = db.CreateConnection();
```

Comment:
```suggestion
            using var connection = db.CreateConnection();
```

The `DbDataSource` does not keep track of handed-out connections; clients must dispose any that are created.

Reply:
Or inject `NpgsqlConnection` directly, as mentioned for MySQL (below).